sdk/src/main/scala/com/horizen/account/state/AccountState.scala

```diff
@@ -400,4 +359,109 @@ object AccountState {
     } else
       throw new RuntimeException("State metadata storage is not empty!")
   }
+
+  def blockGasLimitExceeded(cumGasUsed: BigInteger): Boolean = {
+    // TODO
+    false
+  }
+
+  def blockSizeExceeded(blockSize: Int, txCounter: Int): Boolean = {
+    if (txCounter > SidechainBlockBase.MAX_SIDECHAIN_TXS_NUMBER || blockSize > SidechainBlockBase.MAX_BLOCK_SIZE)
+      true // stop data collection
+    else {
+      false // continue data collection
+    }
+  }
+
+  def applyAndGetReceipts(stateView: AccountStateView,
+                          mainchainBlockReferencesData: Seq[MainchainBlockReferenceData],
+                          sidechainTransactions: Seq[SidechainTypes#SCAT],
+                          inputBlockSize: Int = 0,
+                          blockHashOpt: Option[Array[Byte]] = None): Try[Seq[EthereumReceipt]] = Try {
+
+    // this function can be called when forging a block or when processing a received block
+    // In the former case:
+    //   1. we do not have (and do not need) height and the block hash
+    //   2. we must skip tx not applicable to the current state
+    //   3. we must include tx up to the block gas limit threshold
+    val isForging = blockHashOpt.isEmpty
+
+    val (blockNumber, blockHash) = if (isForging) {
+      (-1, Utils.ZEROS_HASH)
+    } else {
+      (stateView.getHeight + 1, blockHashOpt.get)
+    }
+
+    for(mcBlockRefData <- mainchainBlockReferencesData) {
+      stateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+    }
+
+    val receiptList = new ListBuffer[EthereumReceipt]()
+
+    var cumGasUsed : BigInteger = BigInteger.ZERO
+    var txsCounter: Int = 0
+    var blockSize: Int = inputBlockSize
+
+    for ((tx, txIndex) <- sidechainTransactions.zipWithIndex) {
+
+      stateView.applyTransaction(tx, txIndex, cumGasUsed) match {
+        case Success(consensusDataReceipt) =>
+          val txGasUsed = consensusDataReceipt.cumulativeGasUsed.subtract(cumGasUsed)
+          // update cumulative gas used so far
+          cumGasUsed = consensusDataReceipt.cumulativeGasUsed
+
+          if (isForging) {
+            blockSize = blockSize + tx.bytes.length + 4 // placeholder for Tx length
+            txsCounter += 1
+
+            if (blockSizeExceeded(blockSize, txsCounter))
+              return Success(receiptList)
+          }
+
+          if (blockGasLimitExceeded(cumGasUsed)) {
+            if (isForging) {
+              return Success(receiptList)
+            } else {
+              log.error("Could not apply tx, block gas limit exceeded")
+              throw new IllegalArgumentException("Could not apply tx, block gas limit exceeded")
+            }
+          }
+
+          val ethTx = tx.asInstanceOf[EthereumTransaction]
+
+          val txHash = idToBytes(ethTx.id)
+
+          // The contract address created, if the transaction was a contract creation
+          val contractAddress = if (ethTx.getTo == null) {
```

Comment:
Are you sure ethTx.getTo == null is enough for selecting a tx creating a smart contract? In EthereumTransaction constructor, if to param is null it is changed in "".
Maybe creating a method in EthereumTransaction class for testing if it is a tx creating a smart contract could be useful

Reply:
Should be safe, after checking the EthereumTransaction object creation code the address is null in case of 'to' string null, "", and also "0x"